### PR TITLE
Kirby's Dream Land 3: Implement New Game

### PR DESCRIPTION
**Pull request by Silvris**
## What is this fixing or adding?
Adds Kirby's Dream Land 3 as a supported game to Archipelago.

This also includes related changes to the plando guide and Generate.
## How was this tested?
3 beta tests
10+ apworld releases
Numerous "unsupported" games, in both sync and async environments

## If this makes graphical changes, please attach screenshots.
AP_95592586329425768063_P1_SilvrisKDL3_000
AP_95592586329425768063_P1_SilvrisKDL3_001
image
P1_SilvrisKDL3_qnrCWxWjQVuOUw39PuRMtQ000

### CODE REVIEW

File: worlds/kdl3/Rules.py
Comment (Alchav) [inline worlds/kdl3/Rules.py:20]: The way this works when `open_world` is disabled makes this game explode out into ~36 spheres. This can make generation needlessly slower as well as skew progression balancing. It looks like only 3 levels have logical requirements to complete them, so I don't think it's necessary to make every level logically dependent on an event like this, only those points where a logical requirement exists should be necessary.

File: worlds/kdl3/Rules.py
Comment (Alchav) [inline worlds/kdl3/Rules.py:101]: Could simplify by removing the previous four lines and change this to:
```suggestion
    for abilities, bukisets in EnemyAbilities.enemy_restrictive[1:5]
```

File: Generate.py
Comment (Alchav) [inline Generate.py:481]:
```suggestion
    if ret.game in ("Minecraft", "Ocarina of Time", "Kirby's Dream Land 3"):
```

**Review by Alchav — CHANGES_REQUESTED** [review]

File: worlds/kdl3/Rules.py
Comment (Silvris) [reply worlds/kdl3/Rules.py:20]: When I did rules on can_reach(Location), I was having generation failures because of it (I forget the specific reason, but it had something to do with sweep not accurately finding all locations that were reachable). I'll try to look into it again though.

File: worlds/kdl3/Options.py
Comment (beauxq) [inline worlds/kdl3/Options.py:358]: https://docs.python.org/3/library/typing.html#the-type-of-class-objects
Type annotations should use square brackets `type[Option]` and not call the `type` function.
But `type[Option]` isn't supported in Python 3.8, so in Python 3.8, it should be `typing.Type[Option]`.

Or, since we have a name for this metaclass, it could be `AssembleOptions`.

But `typing.Type[Option]` might be better understood by readers.

Comment (beauxq) [reply worlds/kdl3/Options.py:358]: Actually, since `Option` is `Generic`, a strict type checker would like to see `typing.Type[Option[typing.Any]]`
(or `AssembleOptions`)

File: worlds/kdl3/Rules.py
Comment (Silvris) [reply worlds/kdl3/Rules.py:18]: After investigation, changing the rule to be can_reach(Location) for `open_world` disabled leads to infrequent test failures seemingly caused by an issue addressed by #1764. As such, I'm going to avoid changing this until that PR is merged.

**Review by ScipioWright — APPROVED** [review]
I have read through parts of the "main" ap files -- init, Rules, Regions, mostly. The rest are far outside of what I know. I have played through a couple of games and did not notice issues. Rolled it a bunch of times with various settings and with various games as a test, did not notice issues.

File: worlds/kdl3/Rules.py
Comment (Silvris) [reply worlds/kdl3/Rules.py]: Figured out a solution without needing 1764.

**Review by agilbert1412 — APPROVED** [review]

File: worlds/generic/docs/plando_en.md
Comment (Silvris) [reply worlds/generic/docs/plando_en.md:220]: I figured if we have to edit the file, might as well fix the dead link. This isn't Minecraft documentation though.

File: worlds/kdl3/Client.py
Comment (Silvris) [reply worlds/kdl3/Client.py:268]: Most things are named, but this is actually just masking my item numbering scheme.
0x770001 (Burning) & 0x000070 = 0
0x770010 (Rick) & 0x000070 = 0x10
0x770020 (Heart Star) & 0x000070 = 0x20
0x770040 (Gooey Bag) & 0x000070 = 0x40

File: worlds/kdl3/Rules.py
Comment (agilbert1412) [inline worlds/kdl3/Rules.py:218]: I think this block should be a method

File: worlds/kdl3/docs/setup_en.md
Comment (agilbert1412) [inline worlds/kdl3/docs/setup_en.md:21]: A lot of ROM setup guide give the expected exact name of the rom, I think it would be nice here as well

**Review by agilbert1412 — APPROVED** [review]

File: worlds/kdl3/docs/setup_en.md
Comment (agilbert1412) [reply worlds/kdl3/docs/setup_en.md:21]: Your call, I just mention what comes to my mind :)

Comment (Silvris) [reply worlds/kdl3/docs/setup_en.md:21]: When settings receives the file, it'll copy it to the Archipelago directory with the correct name. I think here we're good to just say "It wants your rom" effectively.

File: worlds/kdl3/data/Rooms.json
Comment (agilbert1412) [inline worlds/kdl3/data/Rooms.json]: Any chance this file could be json-prettified?
This would make it so that changes to it are actually readable, in the code but also in the git diff.

**Review by Berserker66 — APPROVED** [review]

File: worlds/generic/docs/plando_en.md
Comment (Berserker66) [inline worlds/generic/docs/plando_en.md:164]: This now removes the link to the LttP Plando guide.

Comment (Silvris) [reply worlds/generic/docs/plando_en.md:164]: Good catch.

File: worlds/generic/docs/plando_en.md
Comment (Berserker66) [inline worlds/generic/docs/plando_en.md:180]: It now recreates the LttP docs here, rather than referring to its guide.

Comment (Silvris) [reply worlds/generic/docs/plando_en.md:180]: I can definitely trim it some, but I think there should be some LttP examples within the main guide since that will still likely be the most prominent use of boss plando. Maybe just leave off the last LttP example?

File: worlds/kdl3/Names/__init__.py
Comment (alwaysintreble) [inline worlds/kdl3/Names/__init__.py:1]: couldn't comment on this in pycharm but why does this file exist?

File: worlds/LauncherComponents.py
Comment (alwaysintreble) [reply worlds/LauncherComponents.py:91]: While I agree I took a glance at it not too long ago and though sniclient is closer than it used to be, it's still quite a bit of work away from being functional as an apworld

File: worlds/kdl3/Names/__init__.py
Comment (Silvris) [reply worlds/kdl3/Names/__init__.py:1]: It showed up when beauxq ran a linter over the files, so it clearly wanted it to exist. I'm fine with it going.

File: worlds/kdl3/Locations.py
Comment (Silvris) [reply worlds/kdl3/Locations.py:15]: Pretty sure this was copy-pasted from one of the older worlds, so I just never touched it.

File: worlds/kdl3/Regions.py
Comment (alwaysintreble) [reply worlds/kdl3/Regions.py:61]: Ah OK that's fair. It still might be worth considering a different data structure for it down the road so you can get the rooms you need on the fly without having to constantly filter it

File: worlds/kdl3/Regions.py
Comment (Silvris) [reply worlds/kdl3/Regions.py:108]: There's a 7th stage in each level (the boss stage) that we have to handle differently. So we can do the regular `for stage in range(6)` to cover the first 6 stages, and then hit the else to catch the 7th.

Comment (alwaysintreble) [reply worlds/kdl3/Regions.py:108]: Or you can just not have the else and have the code after the for block 😛

File: worlds/kdl3/Rules.py
Comment (Silvris) [reply worlds/kdl3/Rules.py:114]: There was definitely a reason at some point, but I'm not sure it exists anymore.

File: worlds/kdl3/__init__.py
Comment (Silvris) [reply worlds/kdl3/__init__.py:289]: Can't believe I missed this one, there was only 2 in the entire world.

File: worlds/kdl3/__init__.py
Comment (Silvris) [reply worlds/kdl3/__init__.py:216]: The issue here is that I don't want any generic filler, only stars. I will move the list creation out of the comprehension though.

File: worlds/kdl3/__init__.py
Comment (Silvris) [reply worlds/kdl3/__init__.py:74]: Everything here is mutable. `self.rom_name` gets set during generate_output.

File: worlds/kdl3/Names/__init__.py
Comment (beauxq) [reply worlds/kdl3/Names/__init__.py:1]: It tells python tools that the directory `Names` is a module that python can import from.

```python
from .Names.AnimalFriendSpawns import animal_friend_spawns
from .Names.EnemyAbilities import vanilla_enemies, enemy_mapping, enemy_restrictive
from .Names import LocationName
```

The python code might work without it, but some analysis tools will get confused by the imports or module structure.

**Review by alwaysintreble — APPROVED** [review]

**Review by alwaysintreble — APPROVED** [review]

File: worlds/generic/docs/plando_en.md
Comment (Berserker66) [reply worlds/generic/docs/plando_en.md:180]: Well, I suppose this will  be stuck until someone who reviews docs looks at it.

Comment (alwaysintreble) [reply worlds/generic/docs/plando_en.md:180]: didn't mention it in my original review but i did look through this and it seems fine to me. the generic doc should definitely have examples in it and these are reasonable. yes the lttp doc also have boss shuffle examples but these are the only two games that support boss plando so it's still relevant imo.

**Review by ThePhar — APPROVED** [review]

**Pull request merged**
